oreon-11-rp1/sdubby/updateloaderentries.8
.\" Manpage for updateloaderentries.sh
.TH man 8 "17 Oct 2022" "1.0" "updateloaderentries.sh man page"
.SH NAME
updateloaderentries.sh \- Merge /etc/kernel/cmdline options into loader entries
.SH SYNOPSIS
updateloaderentries.sh
.SH DESCRIPTION
updateloaderentries.sh, merges the kernel command line options in /etc/kernel/cmdline
into the /boot/efi/loader/entries/ files utilized by systemd-boot (and others) while
stripping install and anaconda specific boot options.
.SH OPTIONS
The updateloaderentries.sh does not take any options.
.SH SEE ALSO
systemd-boot(7), loader.conf(5)
.SH BUGS
No known bugs.



bsf/changelog
* Wed Jul 17 2024 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-56
- Rebuilt for https://fedoraproject.org/wiki/Fedora_41_Mass_Rebuild

* Fri May 31 2024 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-55
- Switch to a newer patch macro syntax

* Tue Feb 27 2024 Jiri Vanek <jvanek@redhat.com> - 2.4.0-54
- Rebuilt for java-21-openjdk as system jdk

* Tue Feb 27 2024 Jiri Vanek <jvanek@redhat.com> - 2.4.0-53
- bump of release for for java-21-openjdk as system jdk

* Tue Feb 06 2024 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-52
- Enable javadocs

* Tue Jan 23 2024 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-51
- Rebuilt for https://fedoraproject.org/wiki/Fedora_40_Mass_Rebuild

* Fri Jan 19 2024 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-50
- Rebuilt for https://fedoraproject.org/wiki/Fedora_40_Mass_Rebuild

* Fri Sep 01 2023 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-49
- Convert License tag to SPDX format

* Wed Jul 19 2023 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-48
- Rebuilt for https://fedoraproject.org/wiki/Fedora_39_Mass_Rebuild

* Wed Jan 18 2023 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-47
- Rebuilt for https://fedoraproject.org/wiki/Fedora_38_Mass_Rebuild

* Fri Aug 05 2022 Marian Koncek <mkoncek@redhat.com> - 2.4.0-46
- Explicitly specify JVM target version

* Wed Jul 20 2022 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-45
- Rebuilt for https://fedoraproject.org/wiki/Fedora_37_Mass_Rebuild

* Sat Feb 05 2022 Jiri Vanek <jvanek@redhat.com> - 2.4.0-44
- Rebuilt for java-17-openjdk as system jdk

* Wed Jan 19 2022 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-43
- Rebuilt for https://fedoraproject.org/wiki/Fedora_36_Mass_Rebuild

* Tue Nov 02 2021 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-42
- Bump Java compiler source/target levels to 1.7

* Wed Jul 21 2021 Fedora Release Engineering <releng@fedoraproject.org> - 2.4.0-41
- Rebuilt for https://fedoraproject.org/wiki/Fedora_35_Mass_Rebuild

* Tue Jan 26 2021 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-40
- Rebuilt for https://fedoraproject.org/wiki/Fedora_34_Mass_Rebuild

* Fri Aug 28 2020 Fabio Valentini <decathorpe@gmail.com> - 0:2.4.0-39
- Re-enable JavaScript support (RHBZ#1858613).

* Mon Jul 27 2020 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-38
- Rebuilt for https://fedoraproject.org/wiki/Fedora_33_Mass_Rebuild

* Mon Jul 20 2020 Fabio Valentini <decathorpe@gmail.com> - 0:2.4.0-37
- Adaptations for Java 11 compatibility.

* Fri Jul 10 2020 Jiri Vanek <jvanek@redhat.com> - 0:2.4.0-36
- Rebuilt for JDK-11, see https://fedoraproject.org/wiki/Changes/Java11

* Tue Jan 28 2020 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-35
- Rebuilt for https://fedoraproject.org/wiki/Fedora_32_Mass_Rebuild

* Tue Nov 05 2019 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-32
- Mass rebuild for javapackages-tools 201902

* Tue Nov 05 2019 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-32
- Mass rebuild for javapackages-tools 201902

* Wed Aug 14 2019 Fabio Valentini <decathorpe@gmail.com> - 0:2.4.0-34
- Remove BuildRequires: rhino.

* Wed Jul 24 2019 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-33
- Rebuilt for https://fedoraproject.org/wiki/Fedora_31_Mass_Rebuild

* Fri May 24 2019 Mikolaj Izdebski <mizdebsk@redhat.com> - 2.4.0-31
- Mass rebuild for javapackages-tools 201901

* Thu Jan 31 2019 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-32
- Rebuilt for https://fedoraproject.org/wiki/Fedora_30_Mass_Rebuild

* Mon Aug 20 2018 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-31
- Re-add bulid dependency on rhino

* Tue Jul 17 2018 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-30
- Remove bulid dependency on rhino

* Thu Jul 12 2018 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-29
- Rebuilt for https://fedoraproject.org/wiki/Fedora_29_Mass_Rebuild

* Wed Feb 07 2018 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-28
- Rebuilt for https://fedoraproject.org/wiki/Fedora_28_Mass_Rebuild

* Wed Jul 26 2017 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-27
- Rebuilt for https://fedoraproject.org/wiki/Fedora_27_Mass_Rebuild

* Fri Feb 10 2017 Michael Simacek <msimacek@redhat.com> - 0:2.4.0-26
- Add missing BR apache-parent

* Fri Feb 10 2017 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-25
- Rebuilt for https://fedoraproject.org/wiki/Fedora_26_Mass_Rebuild

* Tue Jun 14 2016 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-24
- Cleanup package

* Wed Feb 03 2016 Fedora Release Engineering <releng@fedoraproject.org> - 0:2.4.0-23
- Rebuilt for https://fedoraproject.org/wiki/Fedora_24_Mass_Rebuild

* Tue Jul 14 2015 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-22
- Add build-requires on javapackages-local

* Wed Jun 17 2015 Fedora Release Engineering <rel-eng@lists.fedoraproject.org> - 0:2.4.0-21
- Rebuilt for https://fedoraproject.org/wiki/Fedora_23_Mass_Rebuild

* Thu May 14 2015 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-20
- Disable javadoc doclint

* Sat Jun 07 2014 Fedora Release Engineering <rel-eng@lists.fedoraproject.org> - 0:2.4.0-19
- Rebuilt for https://fedoraproject.org/wiki/Fedora_21_Mass_Rebuild

* Wed May 21 2014 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-18
- Use .mfiles generated during build
- Update to current packaging guidelines

* Sat Aug 03 2013 Fedora Release Engineering <rel-eng@lists.fedoraproject.org> - 0:2.4.0-17
- Rebuilt for https://fedoraproject.org/wiki/Fedora_20_Mass_Rebuild

* Wed Feb 13 2013 Fedora Release Engineering <rel-eng@lists.fedoraproject.org> - 0:2.4.0-16
- Rebuilt for https://fedoraproject.org/wiki/Fedora_19_Mass_Rebuild

* Thu Nov 22 2012 Tomas Radej <tradej@redhat.com> - 0:2.4.0-15
- Fixed URL of Source0

* Tue Nov 20 2012 Mikolaj Izdebski <mizdebsk@redhat.com> - 0:2.4.0-14

oreon-11-rp1/sdubby/updateloaderentries.sh
#!/bin/bash
# this utility should take the boot params from existing
# loader entries files (from anaconda+linux-kernel install)
# which should be a clone of the machine's /proc/cmdline
# and removes the "inst." options "BOOT_IMG="
# and then merges it with the /etc/kernel/cmdline
# which is generated from anaconda
#
# It also acts as a lightweight grubby for
# printing kernel paths, and updating individual bls entries.
# Its largely intended to shim kdumpctl->grubby and anaconda
# any usefulness beyond that is a bonus.
progname=`basename $0`
utilmode="help"
modify_args=""

# sed and awk invocations for manipulating loader entries and bootctl output
BOOTCTL_TO_GRUBBY_INFO="sed -e \"s/ *options: /args:=/g\"  -e \"s/ *linux: /kernel:=/g\" -e \"s/ *initrd: /initrd:=/g\" -e \"s/ *id: /id:=/g\" -e \"s/ *title: /title:=/g\""
BOOTCTL_REMOVE_UNUSED="sed -e \"/source:/d\" -e \"/machine-id:/d\" -e \"/type:/d\" -e \"/sort-key:/d\" -e \"/version:/d\""
BOOTCTL_QUOTING="sed -e \"s/^\(.*\):=\(.*\)/\\1=\\\"\\2\\\"/\""
FIXCMDLINE='
/^options/ {
  for (i=2;i<=NF;i++) {
    if ($i ~ /^BOOT_IMAGE/)
      continue;
    if ($i ~ /^inst\./)
      continue;
    options[$i] = "1"
  }
  # merge /etc/kernel/cmdline
  while (getline<"/etc/kernel/cmdline") {
    for (i=1;i<=NF;i++) {
        options[$i] = "1"
    }
  }
  printf "options    "
  for (i in options)
      printf "%s ",i
  printf "\n"
  next
}
{ print }
'
CAPTUREROOT='
/options: / {
  for (i=1;i<=NF;i++) {
    if ($i ~ /root=/)
        print "root=\"" substr($i,6) "\""
  }
}
{ print }
'

[[ $? = 0 ]] || exit 1

if [[ "$progname" == "updateloaderentries" ]]; then
    utilmode="merge"
fi


OPTS="$(getopt -o hc:i:b:? --long help,args:,info:,update-kernel:,remove-args:,default-kernel -- "$@")"

eval set -- "$OPTS"

while [ ${#} -gt 0 ]; do
    case "$1" in
	--help)
	    utilmode="help"
	    ;;
	--info)
	    info_kernel="$2"
	    utilmode="info"
	    shift 1
	    ;;
	--args)
	    modify_args="$2"
	    utilmode="add"
	    shift 1
	    ;;
	--remove-args)
	    modify_args="$2"
	    utilmode="remove"
	    shift 1
	    ;;
	--update-kernel)
	    info_kernel="$2"
	    shift 1
	    ;;
	--default-kernel)
	    utilmode="defaultkernel"
	    ;;
    esac
    shift
done

if [[ "$info_kernel" == "ALL" ]]; then
    unset info_kernel
fi

case "$utilmode" in
    help)
	echo "updateloaderentries.sh: When run as updateloaderentries it doesn't require any parameters and merges"
	echo "                        The /etc/kernel/cmdline and /proc/cmdline into all systemd-boot configurations."
	echo "grubby: When run symlinked as grubby, it takes the grubby --args, --remove-args and --info"
	echo "        switches. All three take a uname -r compatible option to specify the kernel/initrd"
	echo "        being modified/shown or ALL."
	;;
    remove)
	config_file=`bootctl list |grep -B7 -A2 ".*linux:.*${info_kernel}"|grep "^.*source:" |cut -d: -f2- `
	sed -i -e "s/^options\(.*\) ${modify_args}/options\\1/" $config_file
	;;
    add)
	config_file=`bootctl list |grep -B7 -A2 ".*linux:.*${info_kernel}"|grep "^.*source:" |cut -d: -f2- `
	for fn in $config_file; do
	    grep ${modify_args} $fn >/dev/null
	    if [[ "$?" == "1" ]]; then
		sed -i -e "s/^options    /options    ${modify_args} /" $fn
	    fi
	done
	;;
    info)
	bootctl list | grep -B7 -A2 ".*linux:.*$info_kernel" | awk "$CAPTUREROOT" | eval "${BOOTCTL_REMOVE_UNUSED}" | eval "${BOOTCTL_TO_GRUBBY_INFO}" | eval "${BOOTCTL_QUOTING}"
	;;
    defaultkernel)
	bootctl list |grep -A6 ".*title:.*(default)" | grep "linux:" | cut -c16-
	;;
    merge)
	for i in /boot/efi/loader/entries/*; do
	    if [[ -f "$i" ]]; then
		/usr/bin/awk -i inplace "$FIXCMDLINE" "$i";
	    fi
	done
        
        # just in case, relink if needed
        if ! [[ -e /sbin/installkernel ]]; then
	    /usr/bin/ln -s /usr/bin/kernel-install /sbin/installkernel
        fi
        if ! [[ -e /sbin/grubby ]]; then
            /usr/bin/ln -s /sbin/updateloaderentries /sbin/grubby
        fi
        ;;
esac

